core/client/fs/src/main/java/alluxio/client/file/cache/CacheManager.java

```diff
@@ -270,9 +277,82 @@ default int get(PageId pageId, int pageOffset, int bytesToRead, byte[] buffer, i
    * @param cacheContext cache related context
    * @return number of bytes read, 0 if page is not found, -1 on errors
    */
-  int get(PageId pageId, int pageOffset, int bytesToRead, PageReadTargetBuffer buffer,
+  int get(PageId pageId, int pageOffset, int bytesToRead, ReadTargetBuffer buffer,
       CacheContext cacheContext);
 
+  /**
+   * Reads a part of a page with metrics recording.
+   *
+   * @param pageId page identifier
+   * @param pageOffset offset into the page
+   * @param bytesToRead number of bytes to read in this page
+   * @param buffer destination buffer to write
+   * @param cacheContext cache related context
+   * @param stopwatch stop watch to record metrics
+   * @return number of bytes read, 0 if page is not found, -1 on errors
+   */
+  default int getWithMetrics(PageId pageId, int pageOffset, int bytesToRead,
+      ReadTargetBuffer buffer, CacheContext cacheContext, Stopwatch stopwatch) {
+    stopwatch.reset().start();
+    int bytesRead =
+        get(pageId, pageOffset, bytesToRead, buffer, cacheContext);
+    stopwatch.stop();
+    if (bytesRead > 0) {
+      MetricsSystem.meter(MetricKey.CLIENT_CACHE_BYTES_READ_CACHE.getName()).mark(bytesRead);
+      if (cacheContext != null) {
+        cacheContext.incrementCounter(MetricKey.CLIENT_CACHE_BYTES_READ_CACHE.getMetricName(), BYTE,
+            bytesRead);
+        cacheContext.incrementCounter(
+            MetricKey.CLIENT_CACHE_PAGE_READ_CACHE_TIME_NS.getMetricName(), NANO,
+            stopwatch.elapsed(TimeUnit.NANOSECONDS));
+      }
+    }
+    return bytesRead;
+  }
+
+  /**
+   * Reads a part of a page if the queried page is found in the cache, stores the result in buffer.
+   * Loads the page otherwise.
+   *
+   * @param pageId page identifier
+   * @param pageOffset offset into the page
+   * @param bytesToRead number of bytes to read in this page
+   * @param buffer destination buffer to write
+   * @param cacheContext cache related context
+   * @param stopwatch stop watch to record metrics
+   * @param externalDataSupplier the external data supplier to read a page
+   * @return number of bytes read, 0 if page is not found, -1 on errors
+   */
+  default int getAndLoad(PageId pageId, int pageOffset, int bytesToRead,
+      ReadTargetBuffer buffer, CacheContext cacheContext,
+      Stopwatch stopwatch, Supplier<byte[]> externalDataSupplier) {
+    int bytesRead = getWithMetrics(pageId, pageOffset,
+        bytesToRead, buffer, cacheContext, stopwatch);
+    if (bytesRead > 0) {
+      return bytesRead;
+    }
+    stopwatch.reset().start();
+    byte[] page = externalDataSupplier.get();
+    stopwatch.stop();
+    if (page.length == 0) {
+      return 0;
+    }
+    buffer.writeBytes(page, pageOffset, bytesToRead);
+    MetricsSystem.meter(MetricKey.CLIENT_CACHE_BYTES_REQUESTED_EXTERNAL.getName())
+        .mark(bytesToRead);
+    if (cacheContext != null) {
```

Comment:
How come cacheContext can be `null`?

Reply:
Good catch, 
The cachecontext in the original LocalCacheFileInStream is a little weird..... let me double check what's the correct way